File: sdk/src/main/scala/com/horizen/vrf/VRFPublicKey.scala
package com.horizen.vrf

import java.util

// See https://tools.ietf.org/id/draft-goldbe-vrf-01.html#rfc.section.2 as functions description

// TO DO: add json representation to all VRF related classes

class VRFPublicKey(val key: Array[Byte]) {
  require(key.length == VRFPublicKey.length)

  def verify(slotNumber: Int, nonceBytes: Array[Byte], proof: VRFProof): Boolean = {
    VrfLoader.vrfFunctions.verify(key, slotNumber, nonceBytes, proof.bytes)
    //val messageWithCorrectLength: Array[Byte] = Utils.doubleSHA256Hash(message)

    //val decoded = proof.bytes.map(byte => (byte ^ key.head).toByte)
    //messageWithCorrectLength.sameElements(decoded)
  } // jni call to Rust impl

  // maybe also a method for verifying VRFPublicKey
  def isValid: Boolean = VrfLoader.vrfFunctions.isValid // jni call to Rust impl

  def bytes: Array[Byte] = key

  override def hashCode(): Int = util.Arrays.hashCode(key)

  override def equals(obj: Any): Boolean = {
    obj match {
      case that: VRFPublicKey => this.key sameElements that.key
      case _ =>
        false
    }
  }
}

object VRFPublicKey {
  val length: Int = 32 //sha256HashLen
  def parseBytes(bytes: Array[Byte]): VRFPublicKey = new VRFPublicKey(bytes)
}

File: sdk/src/main/java/com/horizen/vrf/VRFPublicKey.java
package com.horizen.vrf;

import com.horizen.utils.BytesUtils;
import scorex.core.serialization.BytesSerializable;
import scorex.core.serialization.ScorexSerializer;

import java.util.Arrays;

public class VRFPublicKey
    implements BytesSerializable
{

  public static final int KEY_LENGTH = 32;

  private byte[] key;

  public VRFPublicKey(byte[] bytes) {
    if (bytes.length != KEY_LENGTH)
      throw new IllegalArgumentException(String.format("Incorrect public key length, %d expected, %d found", KEY_LENGTH, bytes.length));
    this.key = bytes;
  }

  private native boolean nativeVerify (byte[] key, byte[] message, byte[] proof); // jni call to Rust impl

  boolean verify(byte[] message, byte[] proof) {
    return nativeVerify(this.key, message, proof);
  }

  @Override
  public byte[] bytes() {
    return Arrays.copyOf(this.key, KEY_LENGTH);
  }

  @Override
  public ScorexSerializer serializer() {
    return VRFPublicKeySerializer.getSerializer();
  }

  public static VRFPublicKey parseBytes(byte[] bytes) {
    return new VRFPublicKey(bytes);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    VRFPublicKey that = (VRFPublicKey) o;
    return Arrays.equals(key, that.key);
  }

  @Override
  public int hashCode() {
    return Arrays.hashCode(key);
  }

  @Override
  public String toString() {
    return "VRFPublicKey{" +
            "key=" + BytesUtils.toHexString(key) +
            '}';
  }

  // maybe also a method for verifying VRFPublicKey
  //def isValid: Boolean = ??? // jni call to Rust impl

}

